lib/transport/transport-tls.c

```diff
@@ -34,15 +34,64 @@ typedef struct _LogTransportTLS
 {
   LogTransportSocket super;
   TLSSession *tls_session;
+  gboolean sending_shutdown;
 } LogTransportTLS;
 
+static inline gboolean
+_is_shutdown_sent(gint shutdown_rc)
+{
+  return shutdown_rc >= 0;
+}
+
+static gint
+log_transport_tls_send_shutdown(LogTransportTLS *self)
+{
+  self->sending_shutdown = TRUE;
+  gint shutdown_rc = SSL_shutdown(self->tls_session->ssl);
+
+  if (_is_shutdown_sent(shutdown_rc))
+    {
+      self->sending_shutdown = FALSE;
+      return shutdown_rc;
+    }
+
+  gint ssl_error = SSL_get_error(self->tls_session->ssl, shutdown_rc);
+
+  switch (ssl_error)
+    {
+    case SSL_ERROR_WANT_READ:
+      self->super.super.cond = G_IO_IN;
+      errno = EAGAIN;
+      break;
+    case SSL_ERROR_WANT_WRITE:
+      self->super.super.cond = G_IO_OUT;
+      errno = EAGAIN;
+      break;
+    case SSL_ERROR_SYSCALL:
+      /* errno is set accordingly */
+      break;
+    default:
+      msg_error("SSL error while shutting down stream",
+                tls_context_format_tls_error_tag(self->tls_session->ctx),
+                tls_context_format_location_tag(self->tls_session->ctx));
+      ERR_clear_error();
+      errno = ECONNRESET;
```

Comment:
> Note that SSL_shutdown() must not be called if a previous fatal error has occurred on a connection i.e. if SSL_get_error() has returned SSL_ERROR_SYSCALL or SSL_ERROR_SSL.

https://www.openssl.org/docs/man1.1.1/man3/SSL_shutdown.html

To me, this means we should not try shutdown anymore, so `self->sending_shutdown` needs to be set to `FALSE` here and the syscall branch.

Reply:
Fatal errors are propagated back to the LogProto layer, where the connection will be closed in such cases.

But you are right, the best would be to set everything to a valid final state. Fixed.

Reply:
Thanks!